tests/src/main/java/net/neoforged/neoforge/debug/transfer/VanillaHandlersTests.java

```diff
@@ -0,0 +1,566 @@
+        if (state.getValue(BlockStateProperties.CHISELED_BOOKSHELF_SLOT_3_OCCUPIED)) throw helper.assertionException("Slot 3 should not be occupied");
+        if (state.getValue(BlockStateProperties.CHISELED_BOOKSHELF_SLOT_4_OCCUPIED)) throw helper.assertionException("Slot 4 should not be occupied");
+        if (state.getValue(BlockStateProperties.CHISELED_BOOKSHELF_SLOT_5_OCCUPIED)) throw helper.assertionException("Slot 5 should not be occupied");
+
+        helper.succeed();
+    }
+
+    @GameTest
+    @EmptyTemplate
+    @TestHolder(description = "Tests that shulker boxes cannot be inserted into other shulker boxes.")
+    public static void testShulkerNoInsert(ExtendedGameTestHelper helper) {
+        BlockPos pos = new BlockPos(0, 2, 0);
+        helper.setBlock(pos, Blocks.SHULKER_BOX);
+        ShulkerBoxBlockEntity shulker = helper.getBlockEntity(pos, ShulkerBoxBlockEntity.class);
+        var resourceHandler = new WorldlyContainerWrapper(shulker, null);
+
+        try (var tx = Transaction.open(null)) {
+            if (resourceHandler.insert(ItemResource.of(Items.SHULKER_BOX), 1, tx) > 0) {
+                helper.fail(Component.literal("Expected shulker box to be rejected"), pos);
+            }
+        }
+
+        helper.succeed();
+    }
+
+    /**
+     * {@link Container#canPlaceItem(int, ItemStack)} is supposed to be independent of the stack size.
+     * However, to limit some stackable inputs to a size of 1, brewing stands and furnaces don't follow this rule in all cases.
+     * This test ensures that the Transfer API works around this issue for furnaces.
+     */
+    @GameTest
+    @EmptyTemplate
+    @TestHolder
+    public static void testBadFurnaceIsValid(ExtendedGameTestHelper helper) {
+        BlockPos pos = new BlockPos(0, 1, 0);
+        helper.setBlock(pos, Blocks.FURNACE.defaultBlockState());
+        FurnaceBlockEntity furnace = helper.getBlockEntity(pos, FurnaceBlockEntity.class);
+        var furnaceWrapper = VanillaContainerWrapper.of(furnace);
+
+        try (Transaction tx = Transaction.open(null)) {
+            if (furnaceWrapper.insert(1, ItemResource.of(Items.BUCKET), 2, tx) != 1) {
+                throw helper.assertionException("Exactly 1 bucket should have been inserted");
+            }
+        }
+
+        helper.succeed();
+    }
+
+    /**
+     * Same as {@link #testBadFurnaceIsValid}, but for brewing stands.
+     */
+    @GameTest
+    @EmptyTemplate
+    @TestHolder
+    public static void testBadBrewingStandIsValid(ExtendedGameTestHelper helper) {
+        BlockPos pos = new BlockPos(0, 1, 0);
+        helper.setBlock(pos, Blocks.BREWING_STAND.defaultBlockState());
+        BrewingStandBlockEntity brewingStand = helper.getBlockEntity(pos, BrewingStandBlockEntity.class);
+        var brewingStandWrapper = VanillaContainerWrapper.of(brewingStand);
+
+        try (Transaction tx = Transaction.open(null)) {
```

Comment:
Can we also add a test that tries to insert (at least) three bottles ignoring index, and making sure that they end up with one in each of the bottle slots?